### ISSUE TITLE
Unit tests are segfaulting

### ISSUE THREAD

**Issue by dasmithii** [petertodd/python-bitcoinlib#30]
Running `python -m unittest discover` in `bitcoin/tests` results in a segmentation fault. I'm having trouble tracking down the issue here.

It occurs within `test_wallet.py.Test_CBitcoinSecret`, seemingly due to an issue with `CBitcoinSecret.from_secret_bytes` in `wallet.py`. But that's about all I know.

**Comment by jamespedwards42** [comment]
Just checking in to confirm the issue on a Mac 10.10.2. Was trying to write a simple script to sign messages, the following fails with a segmentation fault:

```python
import bitcoin
from bitcoin.wallet import CBitcoinSecret

bitcoin.SelectParams('mainnet')
seckey = CBitcoinSecret('5KJvsngHeMpm884wtkJNzQGaCErckhHJBGFsvd3VyK5qMZXj3hS')
```

**Comment by petertodd** [comment]
@jamespedwards42 Thanks for the report! Is it possible to get a version of the library with debug symbols on OS X and run it against that? IE with core dumps enabled? A stack trace might help figure this one out.

**Comment by jamespedwards42** [comment]
Sorry, I'm not familiar with python at all.  I went with importing enabling `faulthandler`, hopefully this helps.

```
Fatal Python error: Segmentation fault

Current thread 0x00007fff7c40b300 (most recent call first):
  File "/Library/Python/2.7/site-packages/python_bitcoinlib-0.2.1-py2.7.egg/bitcoin/core/key.py", line 54 in set_secretbytes
  File "/Library/Python/2.7/site-packages/python_bitcoinlib-0.2.1-py2.7.egg/bitcoin/wallet.py", line 206 in __init__
  File "/Library/Python/2.7/site-packages/python_bitcoinlib-0.2.1-py2.7.egg/bitcoin/wallet.py", line 238 in __init__
  File "authDemo.py", line 8 in <module>
[1]    80854 segmentation fault  python authDemo.py
```

**Comment by jamespedwards42** [comment]
Just realized the line numbers don't line up with the latest version in the repo.  I was using the library from https://pypi.python.org/pypi/python-bitcoinlib/0.2.1.... Checked out the repo and re-ran:

```
File "/Users/james/git/python-bitcoinlib/bitcoin/core/key.py", line 61 in set_secretbytes
  File "/Users/james/git/python-bitcoinlib/bitcoin/wallet.py", line 212 in __init__
  File "/Users/james/git/python-bitcoinlib/bitcoin/wallet.py", line 244 in __init__
```

**Comment by posita** [comment]
I can confirm with python 2.7, 3.4, and pypy 2.5.1 (built via MacPorts). It looks like it's a call out to `libssl`:

```py
...
_ssl = ctypes.cdll.LoadLibrary(ctypes.util.find_library('ssl') or 'libeay32')
...
    def set_secretbytes(self, secret):
        priv_key = _ssl.BN_bin2bn(secret, 32, _ssl.BN_new())
...
```

```sh
% uname -a
Darwin ... 13.4.0 Darwin Kernel Version 13.4.0: Wed Mar 18 16:20:14 PDT 2015; root:xnu-2422.115.14~1/RELEASE_X86_64 x86_64
% python3.4 -c 'import ctypes.util ; _ssl = ctypes.cdll.LoadLibrary(ctypes.util.find_library("ssl")) ; print(_ssl) ; _ssl.BN_bin2bn("\xc4\xbb\xcb\x1f\xbe\xc9\x9de\xbfY\xd8\\\x8c\xb6.\xe2\xdb\x96?\x0f\xe1\x06\xf4\x83\xd9\xaf\xa7;\xd4\xe3\x9a\x8a", 32, _ssl.BN_new())'
<CDLL '/usr/lib/libssl.dylib', handle <_rawffi.alt.CDLL object at 0x000000010aa3b600> at 0x10aa2e988>
zsh: segmentation fault  python3.4 -c
```

I originally thought the problem was that `ctypes.util.find_library('ssl')` was finding OS X's old and broken OpenSSL (in `/usr/lib/libssl.dylib`), but the call seg. faults even if you have a more reasonable version installed (e.g., via MacPorts, Homebrew, etc.):

```sh
% python3.4 -c 'import ctypes.util ; _ssl = ctypes.cdll.LoadLibrary("/opt/macports/lib/libssl.dylib") ; print(_ssl) ; _ssl.BN_bin2bn("\xc4\xbb\xcb\x1f\xbe\xc9\x9de\xbfY\xd8\\\x8c\xb6.\xe2\xdb\x96?\x0f\xe1\x06\xf4\x83\xd9\xaf\xa7;\xd4\xe3\x9a\x8a", 32, _ssl.BN_new())'
<CDLL '/opt/macports/lib/libssl.dylib', handle 7fc881457840 at 100e77a58>
zsh: segmentation fault  python3.4 -c
% otool -L /opt/macports/lib/libssl.dylib
/opt/macports/.site/lib/libssl.dylib:
        /opt/macports/.site/lib/libssl.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
        /opt/macports/.site/lib/libcrypto.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
        /opt/macports/.site/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.8)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```

**Comment by posita** [comment]
Failing on the call to `BN_bin2bn` (in case that wasn't obvious):

```sh
% python3.4 -c 'import ctypes.util ; _ssl = ctypes.cdll.LoadLibrary("/opt/macports/lib/libssl.dylib") ; print(_ssl) ; val = b"\x00\x00\x00\x00" ; bn = _ssl.BN_new() ; print(bn) ; _ssl.BN_bin2bn(val, len(val), bn)'
<CDLL '/opt/macports/lib/libssl.dylib', handle 7fae1b718df0 at 1033d5978>
460325856
zsh: segmentation fault  python3.4 -c
```

**Comment by posita** [comment]
Possibly related to #18?

**Comment by posita** [comment]
Other than creating a dependency, is there a reason not to use [pyOpenSSL](/pyca/pyopenssl) as an interface to the various `BN_*` calls? This works:

```sh
% python -c 'from OpenSSL._util import lib as _ssl ; from OpenSSL._util import ffi as _ffi ; bn = _ssl.BN_bin2bn(b"\xc4\xbb\xcb\x1f\xbe\xc9\x9de\xbfY\xd8\\\x8c\xb6.\xe2\xdb\x96?\x0f\xe1\x06\xf4\x83\xd9\xaf\xa7;\xd4\xe3\x9a\x8a", 32, _ssl.BN_new()) ; print(_ffi.string(_ssl.BN_bn2hex(bn)))'
C4BBCB1FBEC99D65BF59D85C8CB62EE2DB963F0FE106F483D9AFA73BD4E39A8A
```

**Comment by posita** [comment]
I took a *brief* look at what it would take to migrate from `ctypes` to using the `cffi.api.FFI` object exposed by [pyca/cryptography](/pyca/cryptography), and the migration appears nontrivial. `cffi` does have some nice convenience features like attaching native destructors to be invoked during garbage collection, e.g.:

```py
new_bn = ffi.gc(_ssl.BN_new(), _ssl.BN_free)
```

However, this project makes a *lot* of calls out to the SSL library (thankfully, they're all in [`key.py`](/petertodd/python-bitcoinlib/blob/master/bitcoin/core/key.py) as far as I can tell). Every call to `_ssl.[...]` would have to be rewritten (or at least reexamined and tested). That's not to say it's a bad idea, just that it was more work than I had originally hoped. (Wishful thinking, I know.) :blush:

**Comment by posita** [comment]
The good news is that I think I spotted at least a couple minor memory leaks:

* [`key.py`#L32](/petertodd/python-bitcoinlib/blob/2a51acf84b2f2470628948efd3e302d15782bd06/bitcoin/core/key.py#L32) - the return value from `_ssl.EC_KEY_new_by_curve_name` is lost, but never freed (happens only once when the module is loaded)
* [`key.py`#L68](/petertodd/python-bitcoinlib/blob/2a51acf84b2f2470628948efd3e302d15782bd06/bitcoin/core/key.py#L68) - `priv_key` is never freed (I might have misunderstood; it's possible the [subsequent call to `_ssl.EC_KEY_set_private_key`](/petertodd/python-bitcoinlib/blob/2a51acf84b2f2470628948efd3e302d15782bd06/bitcoin/core/key.py#L75) passes off responsibility to the `EC_KEY`, but I still think there will be a reference count problem, which would keep the memory allocated for that `BN` around forever)

**Comment by HelloZeroNet** [comment]
Possible fix for the problem:
 - http://pastebin.com/bQtdDzHx segfault on mac
 - http://pastebin.com/H1XikJFd works fine (needs [some modification](https://github.com/Bitmessage/PyBitmessage/blob/master/src/pyelliptic/openssl.py#L301-L305))

We have tested on 2 MacBook and the second one works on both, so it's looks like using ctype's type hinting fixes the crash.

I also [had strange random SIGSEGV](https://gist.github.com/HelloZeroNet/e05b9cc3adc0e3548137) on Linux, not sure if related, but not happened since we applied the fix.

**Comment by posita** [comment]
@HelloZeroNet, are you interested in submitting a PR? (See also @petertodd's [comment](https://github.com/petertodd/python-bitcoinlib/pull/62#issuecomment-104955665) regarding my PR #62.)

**Comment by HelloZeroNet** [comment]
@posita Sorry, I don't have the necessary modifications. We have moved away from python-bitcoinlib because of the crashes and currently using the [simple script I linked before](http://pastebin.com/H1XikJFd). (we only need speedup for ECDSA_SIG_recover_key_GFp, the rest of the bitcoin operations done by https://github.com/vbuterin/pybitcointools)

**Comment by posita** [comment]
I believe this can be marked as fixed since #79 was merged. (My apologies for not including a `fixes #30` or similar comment in my commit message to automatically close this.) @dasmithii can you confirm with `master` since you are the OP?
